src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkCustomCommandLine.java

```diff
@@ -1521,46 +1602,63 @@ public Iterator<String> iterator() {
 
     @Override
     public int numArgs() {
-      return (beforeEach != null ? 2 : 1) * stringValues.size();
+      return (beforeEach != null ? 2 : 1) * values.size();
     }
 
     @Override
-    public int totalArgLength() {
+    public int totalArgLength(PathMapper pathMapper) {
       int total = 0;
-      for (String arg : stringValues) {
-        total += arg.length();
+      for (Object arg : values) {
+        switch (arg) {
+          case String string -> total += string.length();
+          case DerivedArtifact artifact ->
+              total +=
+                  artifact.getExecPathString().length()
+                      + pathMapper.computeExecPathLengthDiff(artifact);
+          default -> throw new AssertionError("Unexpected value type: " + arg);
+        }
       }
       if (formatEach != null) {
-        total += SingleStringArgFormatter.formattedLength(formatEach) * stringValues.size();
+        total += SingleStringArgFormatter.formattedLength(formatEach) * values.size();
       }
       if (beforeEach != null) {
-        total += beforeEach.length() * stringValues.size();
+        total += beforeEach.length() * values.size();
       }
       return total + numArgs();
     }
   }
 
   /** Preprocessed version of a {@link VectorArg} originating from {@code Args.add_joined}. */
   private static final class JoinedPreprocessedVectorArg implements PreprocessedArg {
-    private final List<String> stringValues;
+    private final List<Object /* String | DerivedArtifact */> values;
     @Nullable private final String formatEach;
     private final String joinWith;
     @Nullable private final String formatJoined;
 
     JoinedPreprocessedVectorArg(
-        List<String> stringValues,
+        List<Object /* String | DerivedArtifact */> values,
         @Nullable String formatEach,
         String joinWith,
         @Nullable String formatJoined) {
-      this.stringValues = stringValues;
+      this.values = values;
       this.formatEach = formatEach;
       this.joinWith = joinWith;
       this.formatJoined = formatJoined;
     }
 
     @Override
-    public Iterator<String> iterator() {
-      Iterator<String> it = stringValues.iterator();
+    public Iterator<String> iterator(PathMapper pathMapper) {
+      Iterator<String> it =
+          Iterables.transform(
+                  values,
+                  value ->
```

Comment:
Same transform is done `UnjoinedPreprocessedVectorArg`. Worth sharing a helper method like `pathMap(Object, PathMapper)`? I think it makes sense to not hide the call to `Iterables.transform` in the helper method.